llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -2452,16 +2465,14 @@ bool SIInsertWaitcnts::runOnMachineFunction(MachineFunction &MF) {
 
   unsigned NumVGPRsMax = ST->getAddressableNumVGPRs();
   unsigned NumSGPRsMax = ST->getAddressableNumSGPRs();
-  assert(NumVGPRsMax <= SQ_MAX_PGM_VGPRS);
+  assert(NumVGPRsMax + AGPR_OFFSET <= SQ_MAX_PGM_VGPRS);
   assert(NumSGPRsMax <= SQ_MAX_PGM_SGPRS);
 
   RegisterEncoding Encoding = {};
-  Encoding.VGPR0 =
-      TRI->getEncodingValue(AMDGPU::VGPR0) & AMDGPU::HWEncoding::REG_IDX_MASK;
-  Encoding.VGPRL = Encoding.VGPR0 + NumVGPRsMax - 1;
-  Encoding.SGPR0 =
-      TRI->getEncodingValue(AMDGPU::SGPR0) & AMDGPU::HWEncoding::REG_IDX_MASK;
-  Encoding.SGPRL = Encoding.SGPR0 + NumSGPRsMax - 1;
+  Encoding.VGPR0 = getRegPoint(*ST, AMDGPU::VGPR0, *TRI);
+  Encoding.VGPRL = Encoding.VGPR0 + NumVGPRsMax * 2 - 1;
+  Encoding.SGPR0 = getRegPoint(*ST, AMDGPU::SGPR0, *TRI);
+  Encoding.SGPRL = Encoding.SGPR0 + NumSGPRsMax * 2 - 1;
```

Comment:
I'm proposing to remove all this in #130056, so you will need to rebase if/when that lands.

Reply:
This patch seems conflicting with the https://github.com/llvm/llvm-project/pull/130056 since this patch could set VGPR0/SGPR0 to be non-zero value and use it to calulate the interval pairs.

Should we add the encoding back? also @kosarev since he is the author of this patch

Reply:
I am not sure what `getHWRegIndex` will return. But at least we have 0x200 for lo16 vgpr, and 0x201 for hi16 vgpr. If getRegpoint is not returnning a constant then we still a var to store it right?

I guess I'll I wait for Ivan's comment

Reply:
Done. I just realized we are counting offset to 0 index gprs here. Rebased